fix(UX): Improve search relevance for link field

ankush opened the pull request:
We use `locate` function to rank matches. It's just SQL analogue of `str.find` but the catch is that it returns 0 when the substring match fails. So the worst matches show up on top. 

Fix: instead of sorting `relevance asc` sort `1/relevance desc`. This is a math hack to convert `0` to `null` and avoid using any complex SQL functionality which isn't allowed in DB query.

Closes https://github.com/frappe/frappe/issues/22725



| Before  | After |
| --- | --- |
|  ![image](https://github.com/frappe/frappe/assets/9079960/58bdbe73-17ae-4d4b-9363-57876991988f) |  ![image](https://github.com/frappe/frappe/assets/9079960/27cc2d0f-7f31-4f43-945e-5a0c8eaefc58)    |
|![image](https://github.com/frappe/frappe/assets/9079960/4d5d490c-0b99-417a-bb58-66fa51920baa) | ![image](https://github.com/frappe/frappe/assets/9079960/49a620f8-e983-4e79-ab31-87f0b015b5ee) |

ankush committed:
fix: Improve search relevance for search_link

When `locate` returns 0 it's shown on top instead it should be shown
last or not shown at all.

This is math hack to avoid using any complex SQL functionality which
isn't allowed in DB query.

ankush commented:
Improvement: Relevance can also be considered for non-name field with less weightage in terms of order. 

E.g. name - relevance * 1, first search field - relevance * 0.5 and so on.

bosue commented:
Truly fascinating, particularly how fast, almost instantly you came up with it…

Now it may not yet fix it all, but already seems a substantial improvement:

https://github.com/frappe/frappe/assets/46800703/3309bff5-cd6e-45e4-a40b-f1ecd4673514

“ta” still on top, with “es” being entered, “Español (es)” still missing, but the other items are already there.

Probably a minor touchup missing, and you solved it with ease. Respect…

ankush commented:
> “ta” still on top,

This is just random behaviour. When nothing is entered, score is same for all entries. So last 10 entries are being shown. 

> with “es” being entered, “Español (es)” still missing

This is because there are a lot of languages starting with `es` and `modified` order is breaking ties in reverse order. So last inserted languages with `es` prefix show up first.  


Both are isolated edge case, so I don't think anything should be done about it IMO. 

---

Update: Both of these issues are kinda addressed by giving higher preference to reference count. i.e. when a link is referred more it will show up higher instead of `modified` order. 

So new tie breaking order is:

1. relevance - substring match and how close to start it is.
2. idx - how many times the document is referred in other documents
3. meta sort order - `modified` mostly and rarely anyone changes this.

ankush committed:
fix: Give idx higher preference than meta order

Meta order in most cases is default "modified" which doesn't quite help.

idx is # of times a document is referred to somewhere else, which is
more likely to be relevant.

ankush commented:
Postgres doesn't like `ifnull(alias) desc` but `alias desc` works? :woozy_face:

ankush committed:
fix: Make search_link query postgres compatible
